### PR TITLE
Runtime metrics docs

### PR DESCRIPTION
**Pull request by lachmatt**
## Why

Document runtime metrics configuration.

## What

Addition of `runtime-metrics.md` doc.

## Tests

n/a

### CODE REVIEW

File: dev/envvars.sh [hunk 1]
```diff
@@ -43,8 +43,6 @@ export CORECLR_PROFILER_PATH="${PWD}/tracer/bin/tracer-home/${OPT_DIR}SignalFx.T
 # Configure SFx .NET Tracer 
 export SIGNALFX_DOTNET_TRACER_HOME="${PWD}/tracer/bin/tracer-home"
 export SIGNALFX_VERSION="1.0.0"
-export SIGNALFX_ENDPOINT_URL=${SIGNALFX_ENDPOINT_URL:-"http://localhost:9411/api/v2/spans"}
-export SIGNALFX_METRICS_ENDPOINT_URL=${SIGNALFX_METRICS_ENDPOINT_URL:-"http://localhost:9943/v2/datapoint"}
```
Comment (lachmatt) [inline dev/envvars.sh:47]: I think these are not needed

File: tracer/src/Datadog.Trace/SignalFx/Metrics/SignalFxReporter.cs [hunk 2]
```diff
@@ -69,7 +69,7 @@ public void Send(DataPointUploadMessage msg)
             }
             catch (Exception ex)
             {
-                Log.Error($"Exception sending metrics: {ex.Message}");
+                Log.Error(ex, $"Exception sending metrics: {ex.Message}");
```
Comment (Kielek) [inline tracer/src/Datadog.Trace/SignalFx/Metrics/SignalFxReporter.cs:72]: I do not think that you need pass ex.Message. USually StackTrace includes it

File: docs/internal/runtime-metrics.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+that forwards them to Splunk Observability Cloud. For guidance on how to setup required components of metrics pipeline in the collector, see 
+[receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/signalfxreceiver) and [exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/signalfxexporter) documentation.
+
+## Exporting directly to Splunk Observabilty Cloud
+If you prefer to send metrics directly to Splunk Observability Cloud, configure the following settings:
+
+| Setting | Value | Notes |
+|-|-|-|
+| `SIGNALFX_ACCESS_TOKEN` | *organization access tokens* | (__required__) See [here](https://docs.splunk.com/Observability/admin/authentication-tokens/org-tokens.html) to learn how to obtain one. |
+| `SIGNALFX_REALM` | *ingest realm* | (__recommended__) Set to send telemetry directly to Splunk Observability Cloud. If configured, metrics are sent to `https://ingest.<SIGNALFX_REALM>.signalfx.com/v2/datapoint`. To find your realm, open Splunk Observability Cloud, click Settings, and click on your user name. __NOTE: affects traces endpoint as well.__ |
```
Comment (Kielek) [inline docs/internal/runtime-metrics.md:25]: as I remeber @theletterf preferred different wording for ingest.
@theletterf, could you please check it?

File: docs/internal/runtime-metrics.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+## How to check where are metrics exported?
+
+Check the value of the `metrics_agent_url` key under `TRACER CONFIGURATION`.
+
+## How to verify collector setup?
+
+* Make sure you are running the [Splunk OpenTelemetry Collector](https://github.com/signalfx/splunk-otel-collector)
+* Make sure that the collector is configured correctly to handle metrics.
+* Make sure that an `signalfx` receiver and `signalfx` exporter are configured in the collector.
```
Comment (pellared) [inline docs/internal/runtime-metrics.md:45]: is it not a description of the line above?

```suggestion
Make sure that a `signalfx` receiver and a `signalfx` exporter are configured in the collector.
```

or maybe the previous line should be removed?

File: docs/internal/runtime-metrics.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+to `true` for your .NET process. 
+
+# Metrics exporter settings
+
+## Default configuration
+Metrics are sent to local [Splunk OpenTelemetry Collector](https://github.com/signalfx/splunk-otel-collector)
+that forwards them to Splunk Observability Cloud. For guidance on how to setup required components of metrics pipeline in the collector, see 
+[receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/signalfxreceiver) and [exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/signalfxexporter) documentation.
+
+## Exporting directly to Splunk Observabilty Cloud
```
Comment (pellared) [inline docs/internal/runtime-metrics.md:19]: there should be a blank line after the header (applies to other places)

File: docs/internal/runtime-metrics.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
+# About the .NET runtime metrics
+
+The SignalFx Instrumentation for .NET includes an automatic runtime metrics collection.
```
Comment (pellared) [inline docs/internal/runtime-metrics.md:3]:
```suggestion
The SignalFx Instrumentation for .NET is able to collect runtime metrics.
```

File: docs/internal/runtime-metrics.md [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+This feature can be enabled with configuration setting and is disabled by default.
```
Comment (pellared) [inline docs/internal/runtime-metrics.md:4]:
```suggestion
This feature can be enabled using configuration settings.
```

File: docs/internal/runtime-metrics.md [hunk 8]
```diff
@@ -0,0 +1,48 @@
+# About the .NET runtime metrics
+
+The SignalFx Instrumentation for .NET includes an automatic runtime metrics collection.
+This feature can be enabled with configuration setting and is disabled by default. 
+When enabled, metrics are periodically captured and sent to Splunk Observability Cloud.
```
Comment (pellared) [inline docs/internal/runtime-metrics.md:5]:
```suggestion
Metrics are exported on a periodic basis (push model).
```

File: docs/internal/runtime-metrics.md (hunk 7 above)
Comment (pellared) [reply docs/internal/runtime-metrics.md:4]: Can we configure the interval?

File: docs/internal/runtime-metrics.md [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+# Metrics exporter settings
+
+## Default configuration
+Metrics are sent to local [Splunk OpenTelemetry Collector](https://github.com/signalfx/splunk-otel-collector)
+that forwards them to Splunk Observability Cloud. For guidance on how to setup required components of metrics pipeline in the collector, see 
+[receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/signalfxreceiver) and [exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/signalfxexporter) documentation.
+
+## Exporting directly to Splunk Observabilty Cloud
+If you prefer to send metrics directly to Splunk Observability Cloud, configure the following settings:
```
Comment (pellared) [inline docs/internal/runtime-metrics.md:20]: I find it hard to process that two possibilities are described at the same time. @theletterf WDYT?

**Review by pellared — COMMENTED** [review]
nit comments. I suggest to use VS Code and use `markdownlint` extension

File: tracer/src/Datadog.Trace/SignalFx/Metrics/SignalFxReporter.cs (hunk 2 above)
Comment (pellared) [reply tracer/src/Datadog.Trace/SignalFx/Metrics/SignalFxReporter.cs:72]: It does not

File: docs/internal/runtime-metrics.md (hunk 7 above)
Comment (lachmatt) [reply docs/internal/runtime-metrics.md:6]: its fixed

File: docs/internal/runtime-metrics.md (hunk 4 above)
Comment (lachmatt) [reply docs/internal/runtime-metrics.md:48]: previous line removed

File: docs/internal/runtime-metrics.md (hunk 9 above)
Comment (theletterf) [reply docs/internal/runtime-metrics.md:24]: You mean Collector + Direct ingest? We may want to clarify when direct ingest should be used first (for example, when installing a collector is not possible). How would you prefer this?

File: docs/internal/runtime-metrics.md [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+that forwards them to Splunk Observability Cloud. For guidance on how to setup required components of metrics pipeline in the collector, see 
+[receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/signalfxreceiver) and [exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/signalfxexporter) documentation.
+
+### Exporting directly to Splunk Observabilty Cloud
+
+If you prefer to send metrics directly to Splunk Observability Cloud, configure the following settings:
+
+| Setting | Value | Notes |
+|-|-|-|
+| `SIGNALFX_ACCESS_TOKEN` | *organization access tokens* | (__required__) See [here](https://docs.splunk.com/Observability/admin/authentication-tokens/org-tokens.html) to learn how to obtain one. |
```
Comment (theletterf) [inline docs/internal/runtime-metrics.md:28]:
```suggestion
| `SIGNALFX_ACCESS_TOKEN` | Your Observability Cloud token | (__required__) See [here](https://docs.splunk.com/Observability/admin/authentication-tokens/org-tokens.html) to learn how to obtain one. |
```

File: docs/internal/runtime-metrics.md [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+| `SIGNALFX_REALM` | *ingest realm* | (__recommended__) Set to send telemetry directly to Splunk Observability Cloud. If configured, metrics are sent to `https://ingest.<SIGNALFX_REALM>.signalfx.com/v2/datapoint`. To find your realm, open Splunk Observability Cloud, click Settings, and click on your user name. __NOTE: affects traces endpoint as well.__ |
```
Comment (theletterf) [inline docs/internal/runtime-metrics.md:29]:
```suggestion
| `SIGNALFX_REALM` | Your Observability Cloud realm | Set to send telemetry directly to Splunk Observability Cloud. If configured, metrics are sent to `https://ingest.<SIGNALFX_REALM>.signalfx.com/v2/datapoint`. To find your realm, open Splunk Observability Cloud, click Settings, then click on your username. Setting this variable also modify the default tracing endpoint. |
```

File: docs/internal/runtime-metrics.md [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+| `SIGNALFX_METRICS_ENDPOINT_URL` | *valid endpoint url* | (__optional__) Overrides other configuration settings for metrics ingestion endpoint. |
```
Comment (theletterf) [inline docs/internal/runtime-metrics.md:30]:
```suggestion
| `SIGNALFX_METRICS_ENDPOINT_URL` | Metric endpoint | (Optional) Overrides other settings for the metrics ingestion endpoint. |
```

File: docs/internal/runtime-metrics.md [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+
+__IMPORTANT:__ One of `SIGNALFX_REALM` or `SIGNALFX_METRICS_ENDPOINT_URL` has to be configured in addition to `SIGNALFX_ACCESS_TOKEN`.
```
Comment (theletterf) [inline docs/internal/runtime-metrics.md:32]:
```suggestion
__IMPORTANT:__ To export data directly to Splunk Observability Cloud, set `SIGNALFX_REALM` or `SIGNALFX_METRICS_ENDPOINT_URL` in addition to `SIGNALFX_ACCESS_TOKEN`.
```

File: docs/internal/runtime-metrics.md [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+
+## Troubleshooting
+
+### How to check if runtime metrics are enabled?
```
Comment (theletterf) [inline docs/internal/runtime-metrics.md:36]:
```suggestion
### Check if runtime metrics are enabled
```

File: docs/internal/runtime-metrics.md [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+
+The SignalFx Instrumentation for .NET logs the profiling configuration at `INF` during the startup.
```
Comment (theletterf) [inline docs/internal/runtime-metrics.md:38]:
```suggestion
The SignalFx Instrumentation for .NET logs the profiling configuration using `INF` log messages during startup. 
```

File: docs/internal/runtime-metrics.md [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+| `SIGNALFX_METRICS_ENDPOINT_URL` | *valid endpoint url* | (__optional__) Overrides other configuration settings for metrics ingestion endpoint. |
+
+__IMPORTANT:__ One of `SIGNALFX_REALM` or `SIGNALFX_METRICS_ENDPOINT_URL` has to be configured in addition to `SIGNALFX_ACCESS_TOKEN`.
+
+## Troubleshooting
+
+### How to check if runtime metrics are enabled?
+
+The SignalFx Instrumentation for .NET logs the profiling configuration at `INF` during the startup. 
+Verify that `runtime_metrics_enabled` key is set to `true` under `TRACER CONFIGURATION`.
```
Comment (theletterf) [inline docs/internal/runtime-metrics.md:39]:
```suggestion
Verify that the `runtime_metrics_enabled` key is set to `true` under `TRACER CONFIGURATION`.
```

File: docs/internal/runtime-metrics.md [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+
+### How to check where are metrics exported?
```
Comment (theletterf) [inline docs/internal/runtime-metrics.md:41]:
```suggestion
### Check where the metrics are being exported
```

File: docs/internal/runtime-metrics.md [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+
+Check the value of the `metrics_agent_url` key under `TRACER CONFIGURATION`.
```
Comment (theletterf) [inline docs/internal/runtime-metrics.md:43]: What does it look like if the export is done locally? (Not sure if it's possible)

File: docs/internal/runtime-metrics.md [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+
+### How to verify collector setup?
```
Comment (theletterf) [inline docs/internal/runtime-metrics.md:45]:
```suggestion
### Verify the Collector setup
```

File: docs/internal/runtime-metrics.md [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+
+* Make sure you are running the [Splunk OpenTelemetry Collector](https://github.com/signalfx/splunk-otel-collector)
```
Comment (theletterf) [inline docs/internal/runtime-metrics.md:47]:
```suggestion
* Make sure that the [Splunk OpenTelemetry Collector](https://github.com/signalfx/splunk-otel-collector) is running.
```

File: docs/internal/runtime-metrics.md [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+* Make sure that an `signalfx` receiver and `signalfx` exporter are configured in the collector.
```
Comment (theletterf) [inline docs/internal/runtime-metrics.md:48]:
```suggestion
* Make sure that a `signalfx` receiver and a `signalfx` exporter are configured in the Collector.
```

File: docs/internal/runtime-metrics.md [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+Ensure that the `access_token` and `realm` fields are [correctly configured](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/signalfxexporter#metrics-configuration).
+Lastly, double check that the metrics pipeline is configured to use
```
Comment (theletterf) [inline docs/internal/runtime-metrics.md:50]:
```suggestion
* Check that the metrics pipeline is configured to use
```

File: docs/internal/runtime-metrics.md [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+
+### How to check where are metrics exported?
+
+Check the value of the `metrics_agent_url` key under `TRACER CONFIGURATION`.
+
+### How to verify collector setup?
+
+* Make sure you are running the [Splunk OpenTelemetry Collector](https://github.com/signalfx/splunk-otel-collector)
+* Make sure that an `signalfx` receiver and `signalfx` exporter are configured in the collector.
+Ensure that the `access_token` and `realm` fields are [correctly configured](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/signalfxexporter#metrics-configuration).
```
Comment (theletterf) [inline docs/internal/runtime-metrics.md:49]:
```suggestion
* Make sure that the `access_token` and `realm` fields are [configured](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/signalfxexporter#metrics-configuration).
```

File: tracer/src/Datadog.Trace/SignalFx/Metrics/SignalFxReporter.cs [hunk 24]
```diff
@@ -69,7 +69,7 @@ public void Send(DataPointUploadMessage msg)
             }
             catch (Exception ex)
             {
-                Log.Error($"Exception sending metrics: {ex.Message}");
+                Log.Error(ex, "An error occurred while sending metrics.");
```
Comment (theletterf) [inline tracer/src/Datadog.Trace/SignalFx/Metrics/SignalFxReporter.cs:72]:
```suggestion
                Log.Error(ex, "Metric data could not be sent.");
```

Can we make this error more informative? What does `{ex.Message}` contains? Can we add perhaps the endpoint URL in the error?

**Review by theletterf — CHANGES_REQUESTED** [review]

File: docs/internal/runtime-metrics.md (hunk 18 above)
Comment (lachmatt) [reply docs/internal/runtime-metrics.md:41]: it looks like this `http://localhost:9943/v2/datapoint`

File: docs/internal/runtime-metrics.md (hunk 9 above)
Comment (pellared) [reply docs/internal/runtime-metrics.md:22]: I mean that it describes that direct ingest can be configured by specifying
- SIGNALFX_ACCESS_TOKEN and SIGNALFX_REALM
OR
- SIGNALFX_ACCESS_TOKEN and SIGNALFX_METRICS_ENDPOINT_URL

File: docs/internal/runtime-metrics.md (hunk 8 above)
Comment (pellared) [reply docs/internal/runtime-metrics.md:5]: I propose the change as they can be sent to the collector.
maybe

```suggestion
When enabled, metrics are periodically captured and sent to the configured endpoint.
```

Comment (theletterf) [reply docs/internal/runtime-metrics.md:5]: +1 Although from a customer perspective, the end goal is getting them to O11y Cloud, either via Collector or directly.

File: docs/internal/runtime-metrics.md [hunk 25]
```diff
@@ -0,0 +1,49 @@
...
+
+### Default configuration
+
+Metrics are sent to the local instance of [Splunk OpenTelemetry Collector](https://github.com/signalfx/splunk-otel-collector),
+which forwards telemetry to Splunk Observability Cloud. For instructions on how to set up required components of the metrics pipeline in the collector, see 
+[receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/signalfxreceiver) and [exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/signalfxexporter) documentation.
+
+### Exporting directly to Splunk Observability Cloud
+
+If you prefer to send metrics directly to Splunk Observability Cloud, configure the following settings:
```
Comment (theletterf) [inline docs/internal/runtime-metrics.md:22]:
```suggestion
If you prefer to send metrics to Splunk Observability Cloud without using the Collector, configure the following settings:
```

File: docs/internal/runtime-metrics.md (hunk 9 above)
Comment (theletterf) [reply docs/internal/runtime-metrics.md:22]: Made an edit in the IMPORTANT note below for that.

**Review by theletterf — APPROVED** [review]

File: tracer/src/Datadog.Trace/SignalFx/Metrics/SignalFxReporter.cs [hunk 26]
```diff
@@ -69,7 +69,7 @@ public void Send(DataPointUploadMessage msg)
             }
             catch (Exception ex)
             {
-                Log.Error($"Exception sending metrics: {ex.Message}");
+                Log.Error(ex, "Metric data could not be sent.");
```
Comment (theletterf) [inline tracer/src/Datadog.Trace/SignalFx/Metrics/SignalFxReporter.cs:72]: @pellared What did {ex.Message} contain? Can we add perhaps the endpoint URL in the error?

Comment (lachmatt) [reply tracer/src/Datadog.Trace/SignalFx/Metrics/SignalFxReporter.cs:72]: I will add the endpoint URL in separate PR

File: docs/internal/runtime-metrics.md (hunk 8 above)
Comment (lachmatt) [reply docs/internal/runtime-metrics.md:5]: leaving it as it is for now

**Review by pellared — APPROVED** [review]

**Review by pjanotti — APPROVED** [review]

**Pull request merged**